### PR TITLE
fix: Disable sandboxing for cypress_test rules

### PR DESCRIPTION
**Pull request by mrmeku**
Cypress has moved away from supporting the env
variable RUN_ELECTRON_AS_NODE which makes it
hard to keep up with their electron server's
sandboxing violations. We now disable the sandbox
so that user's can don't run into these issues.

---

### Type of change

- Bug fix (change which fixes an issue)

### Test plan
- New test cases added
(using a cypress version 13.6.6 required this change on macOS)

### CODE REVIEW

File: cypress/defs.bzl [hunk 1]
```diff
@@ -12,6 +12,30 @@ _cypress_test = rule(
...
+        chdir = native.package_name(),
+        enable_runfiles = select({
+            "@aspect_rules_js//js/private:enable_runfiles": True,
+            "//conditions:default": False,
+        }),
+        unresolved_symlinks_enabled = select({
+            "@aspect_rules_js//js/private:experimental_allow_unresolved_symlinks": True,
+            "//conditions:default": False,
+        }),
+        # Turn off sandboxing by default to allow electron to perform write
```
Comment (jbedard) [inline cypress/defs.bzl:31]: What type of writing does it do? Is not now going to be modifying files in the bindir? Or runfiles?

File: e2e/workspace/MODULE.bazel [hunk 2]
```diff
@@ -10,6 +10,7 @@ npm = use_extension("@aspect_rules_js//npm:extensions.bzl", "npm", dev_dependenc
 npm.npm_translate_lock(
     name = "npm",
     pnpm_lock = "//:pnpm-lock.yaml",
+    run_lifecycle_hooks = False,
```
Comment (jbedard) [inline e2e/workspace/MODULE.bazel:13]: Are all these minor changes just simplifying tests and formatting things?

Comment (mrmeku) [reply e2e/workspace/MODULE.bazel:13]: Bingo, just to make the example a bit more performant

File: cypress/defs.bzl (hunk 1 above)
Comment (mrmeku) [reply cypress/defs.bzl:31]: No, just files in the electron `user data` folder. Cypress doesn't expose us the electron hook we need to customize this folder from its defaults

```
appData Per-user application data directory, which by default points to:
%APPDATA% on Windows
$XDG_CONFIG_HOME or ~/.config on Linux
~/Library/Application Support on macOS
```
See https://www.electronjs.org/docs/latest/api/app

We set `XDG_CONFIG_HOME` on linux but on Mac we can't fix the sandboxing violation at this time without upstream changes to cypress

Comment (matthewjh) [reply cypress/defs.bzl:31]: Just an idea but did you try setting `HOME`? Maybe you can make it write to `/some/safe/path/Library/Application Support`.

File: e2e/workspace/MODULE.bazel (hunk 2 above)
Comment (matthewjh) [reply e2e/workspace/MODULE.bazel:13]: Why disable all lifecycle hooks? Seems like this might make the e2e test suppress errors that users will see, unless they too set `run_lifecycle_hooks = False` globally, which they probably would not and should not do.

Comment (matthewjh) [reply e2e/workspace/MODULE.bazel:13]: as a user, I did add

```python
lifecycle_hooks_envs = {
        "cypress": ["CYPRESS_INSTALL_BINARY=0"],
    },
```

to `npm_translate_lock`, which is probably better for your e2e test than a global disable (assuming one of the performance issues you allude to is cypress downloading its binary files).

File: cypress/defs.bzl (hunk 1 above)
Comment (mrmeku) [reply cypress/defs.bzl:31]: Yeah, the existing rule sets home already. Unfortunately, the electron process seems to ignore this

File: e2e/workspace/MODULE.bazel (hunk 2 above)
Comment (mrmeku) [reply e2e/workspace/MODULE.bazel:13]: Oh that's awesome! I'll mirror that here and make a comment in the readme!

File: e2e/workspace/module_test/runner.js [hunk 3]
```diff
@@ -12,7 +12,7 @@ async function main() {
     return 1;
   }
 
-  if (result.status !== "finished") {
```
Comment (gregmagolan) [inline e2e/workspace/module_test/runner.js:15]: what's the story here? status used to be reported "finished"?

File: cypress/defs.bzl (hunk 1 above)
Comment (gregmagolan) [reply cypress/defs.bzl:31]: Can you add to the comment an example error message that occurs when sandboxed with newer versions of cypress?

File: e2e/workspace/module_test/runner.js (hunk 3 above)
Comment (mrmeku) [reply e2e/workspace/module_test/runner.js:15]: Yeah, that changed between cypress versions. I’ll leave a comment

File: cypress/defs.bzl [hunk 4]
```diff
@@ -12,6 +12,30 @@ _cypress_test = rule(
...
+        # operations. Cypress does not expose the underlying electron apis so we
+        # cannot alter the user app data directory to be within the bazel
+        # sandbox.
+        tags = kwargs.pop("tags", []) + ["no-sandbox"],
```
Comment (gregmagolan) [inline cypress/defs.bzl:35]: Add an escape valve for users to enable the sandbox if they want to? Maybe they are on an older version of cypress where it works fine.

Sometimes I'll make the escape valve an undocumented attribute such as,
```
tags = kwargs.pop("tags", [])
if kwargs.pop("allow_sandbox", False):
  tags.append("no-sandbox")
```

File: e2e/workspace/MODULE.bazel (hunk 2 above)
Comment (gregmagolan) [reply e2e/workspace/MODULE.bazel:13]: Add this to this PR and add the comment to the README in this PR or a follow-up?

File: e2e/workspace/cli_test/cypress.config.ts [hunk 5]
```diff
@@ -1,13 +1,13 @@
-import { defineConfig } from 'cypress'
+import { defineConfig } from "cypress";
```
Comment (gregmagolan) [inline e2e/workspace/cli_test/cypress.config.ts:1]: Prettier?

File: cypress/private/cypress_test.bzl [hunk 6]
```diff
@@ -4,6 +4,25 @@ load("@aspect_rules_js//js:libs.bzl", "js_binary_lib", "js_lib_helpers")
 load("@bazel_skylib//lib:dicts.bzl", "dicts")
 
 _attrs = dicts.add(js_binary_lib.attrs, {
+    "allow_sandbox": attr.bool(
```
Comment (gregmagolan) [inline cypress/private/cypress_test.bzl:7]: Docstring here is great but think it is misplaced here since `allow_sandbox` is a hidden kwargs parameter on the macro and not an attribute of the rule.

You could put this comment in the macro:

```
    if not kwargs.pop("allow_sandbox", False):
        # comment why where are doing this
        tags.append("no-sandbox")
```

or else hoist `allow_sandbox` do an explicit parameter in the macro and add it to the macro docstring

File: MODULE.bazel [hunk 7]
```diff
@@ -6,12 +6,6 @@ module(
     compatibility_level = 1,
 )
 
-cypress = use_extension("//cypress:extensions.bzl", "cypress")
```
Comment (gregmagolan) [inline MODULE.bazel:9]: We want to punt the toolchain registration to user space so they can pick their cypress version right?

Comment (gregmagolan) [reply MODULE.bazel:9]: Actually, that reminds me that we don't have the MODULE.bazel snipped in our release notes: 

https://github.com/aspect-build/rules_cypress/releases/tag/v0.4.0

like we do for some of other rulesets. Its a mixed bag

bzlmod snippets: rules_esbuild, rules_js, rules_swc, rules_ts
no bzlmod snippets: rules_cypress, rules_jasmine, rules_jest, rules_rollup, rules_terser, rules_webpack

File: e2e/workspace/MODULE.bazel [hunk 8]
```diff
@@ -4,12 +4,22 @@ local_path_override(
...
+register_toolchains("@cypress_toolchains//:all")
+
 bazel_dep(name = "aspect_rules_js", version = "1.41.2", dev_dependency = True)
 bazel_dep(name = "bazel_features", version = "1.9.0", dev_dependency = True)
 
 npm = use_extension("@aspect_rules_js//npm:extensions.bzl", "npm", dev_dependency = True)
 npm.npm_translate_lock(
     name = "npm",
+    # Speed up installation by disabling cypress binary install. Optional.
+    lifecycle_hooks_envs = {
```
Comment (gregmagolan) [inline e2e/workspace/MODULE.bazel:20]: Lets pre-factor all of the things in this PR that are not related to the "no-sandbox" change :)

This one, the prettier changes and "short" tags, can just go in a "cleanup" PR.

File: e2e/workspace/module_test/runner.js (hunk 3 above)
Comment (gregmagolan) [reply e2e/workspace/module_test/runner.js:15]: Does this imply we have some minimum cypress version that will work with the rule set?

Lets also move this to a separate PR since its not related to the "no-sandbox" tag

File: MODULE.bazel (hunk 7 above)
Comment (mrmeku) [reply MODULE.bazel:9]: Bingo, that was an oversite on my part. Its moved now. Lets add the bzlmod snippet either manually or in a seperate PR

Comment (gregmagolan) [reply MODULE.bazel:9]: Yeah, bzlmod snipped can be in a follow-up 👍 This is not the only rule set that is missing one.

File: cypress/private/cypress_test.bzl (hunk 6 above)
Comment (gregmagolan) [reply cypress/private/cypress_test.bzl:7]: this attr should just be removed entirely since it is not used at all right?

File: cypress/defs.bzl [hunk 9]
```diff
@@ -82,23 +133,39 @@ def cypress_module_test(name, runner, cypress = "//:node_modules/cypress", **kwa
         runner: JS file to call into the cypress module api
             See https://docs.cypress.io/guides/guides/module-api
         cypress: The cypress npm package which was already linked using an API like npm_link_all_packages.
+        allow_sandbox: Turn off sandboxing by default to allow electron to perform write operations.
```
[the remark this answers is not in the document]
Comment (mrmeku) [reply cypress/defs.bzl]: I'll switch it to `disable_sandbox` and default to True

**Comment by mrmeku** [comment]
* **#66** <a href="https://app.graphite.dev/github/pr/aspect-build/rules_cypress/66?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @mrmeku and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by gregmagolan — APPROVED** [review]
